### PR TITLE
Adds more tests and fixes #2

### LINKED ISSUE #2: exo-clone does not work if a service's local field points to a directory above the app directory

### PR DESCRIPTION
**Pull request by martinjaime**
@kevgo

### CODE REVIEW

File: features/clone.feature [hunk 1]
```diff
@@ -33,3 +33,53 @@ Feature: cloning an Exosphere application and its services
       | application.yml | app       |
       | service.yml     | app/web   |
       | service.yml     | app/users |
+
+  Scenario: clone an application with services that should reside above application directory.
```
Comment (kevgo) [inline features/clone.feature]: Remove the word "should"

File: features/clone.feature [hunk 2]
```diff
@@ -33,3 +33,53 @@ Feature: cloning an Exosphere application and its services
...
+    Given I am in an empty folder
+    And source control contains the services "web" and "users"
+    And source control contains a repo "app" with a file "application.yml" and the content:
+      """
+      name: Example application
+      description: Demonstrates basic Exosphere application startup
```
Comment (kevgo) [inline features/clone.feature]: This description doesn't seem to fit what the scenario describes

File: features/clone.feature [hunk 3]
```diff
@@ -33,3 +33,53 @@ Feature: cloning an Exosphere application and its services
...
+      version: '1.0'
+
+      services:
+        web:
+          local: ../web
+          origin: ../origins/web
+        users:
+          local: ./users
```
Comment (kevgo) [inline features/clone.feature]: Do you even need this service in this app? I think there is another scenario that describes how internal services are cloned, so no need to repeat this here.

File: features/clone.feature [hunk 4]
```diff
@@ -33,3 +33,53 @@ Feature: cloning an Exosphere application and its services
...
+          origin: ../origins/users
+      """
+    When running "exo-clone origins/app" in the "tmp" directory
```
Comment (kevgo) [inline features/clone.feature]: No need to say "in the tmp directory". This is implicit. All commands happen there.

File: features/clone.feature [hunk 5]
```diff
@@ -33,3 +33,53 @@ Feature: cloning an Exosphere application and its services
...
+    Then it creates the files:
+      | FILE            | FOLDER    |
+      | application.yml | app       |
+      | service.yml     | web       |
+      | service.yml     | app/users |
+
```
Comment (kevgo) [inline features/clone.feature]: Please do 2 empty lines between scenarios, so that they are visually separated from each other. Scenarios can  contain a single empty line inside.

File: features/clone.feature [hunk 6]
```diff
@@ -33,3 +33,53 @@ Feature: cloning an Exosphere application and its services
...
+    Given I am in an empty folder
+    And source control contains the services "web" and "users"
+    And source control contains a repo "app" with a file "application.yml" and the content:
+      """
+      name: Example application
+      description: Demonstrates basic Exosphere application startup
+      version: '1.0'
+
+      services:
+        foo:
```
Comment (kevgo) [inline features/clone.feature]: Let's call this service "nonexisting"

File: features/clone.feature [hunk 7]
```diff
@@ -33,3 +33,53 @@ Feature: cloning an Exosphere application and its services
...
+          local: ./foo
+          origin: ../origins/foo
+        web:
```
Comment (kevgo) [inline features/clone.feature]: I don't think you need the "web" and "users" service in this scenario at all.

File: features/clone.feature [hunk 8]
```diff
@@ -33,3 +33,53 @@ Feature: cloning an Exosphere application and its services
...
+          origin: ../origins/web
+        users:
+          local: ./users
+          origin: ../origins/users
+      """
+    When running "exo-clone origins/app" in the "tmp" directory
+    Then it prints "fatal: repository '../origins/foo' does not exist" in the terminal
+    And it prints "web  done." in the terminal
+    And it prints "users  done." in the terminal
+    And it prints "exo-clone  Done" in the terminal
```
Comment (kevgo) [inline features/clone.feature]: Let's also describe what files get created. I think the right behavior here is to not leave a half-finished app on the hard drive. Let's delete all files that were created!

File: src/app-cloner.ls [hunk 9]
```diff
@@ -24,10 +24,10 @@ class AppCloner extends EventEmitter
       @app-config = yaml.safe-load fs.read-file-sync(config-path, 'utf8')
       @emit 'app-config-ready', @app-config
       cloners = for service-name in Object.keys @app-config.services
-        service-root = path.join @repository.path, path.dirname @app-config.services[service-name].local
+        service-root = path.join @repository.path
```
Comment (kevgo) [inline src/app-cloner.ls]: No need to call `path.join` if you only give it one argument  :)

File: src/app-cloner.ls [hunk 10]
```diff
@@ -24,10 +24,10 @@ class AppCloner extends EventEmitter
...
         service-dir = path.join @repository.path, @app-config.services[service-name].local
         service-origin = @app-config.services[service-name].origin
-        new ServiceCloner service-name, root: service-root, path: service-dir, origin: service-origin
+        new ServiceCloner service-name, cwd: service-root, path: service-dir, origin: service-origin
```
Comment (kevgo) [inline src/app-cloner.ls]: Why do you rename this? `cwd` stands for "current working directory", i.e. the directory from which the user called the exo command. This parameter doesn't tell the ServiceCloner what the current working directory is. It tells it which directory to put the new service into. I think "root" or "root-dir" is a more appropriate name here.

**Comment by kevgo** [comment]
Cool, thanks for doing this! Can you please also create a scenario for cloning an app when one of the directories to clone into already exists? In this case it should abort the clone.

File: src/app-cloner.ls (hunk 10 above)
Comment (martinjaime) [inline src/app-cloner.ls]: I had changed it to `cwd` since it was where the clone command was going to be run from. This is because I want to reference the target location of the cloned repo with reference from the root of the app. Not the root of the service.

Comment (kevgo) [inline src/app-cloner.ls]: "root" here means the root of the service, not the root of the app. If a service would want to know about the root of an app, that parameter would be called `app-root`. 

The word `cwd` usually means the current directory from which the program was started from. If you call `process.cwd`, you get that directory. We shouldn't overload that meaning. 

So, please change this back to `root` here.

**Comment by martinjaime** [comment]
@kevgo, most recent comments addressed. Two more scenarios added.

File: features/clone.feature [hunk 11]
```diff
@@ -27,9 +27,99 @@ Feature: cloning an Exosphere application and its services
...
+          local: ./users
+          origin: ../origins/users
+      """
+    And the "app" clone already exists in the workspace
+    When running "exo-clone origins/app" in the terminal
+    Then it prints "exo-clone  Error: cloning app failed" in the terminal
+    Then I am in an empty folder
+
+
+  Scenario: Cloning an application that depends on a service clone that already exists.
```
Comment (kevgo) [inline features/clone.feature]: Scenario: a service directory already exists

File: features/clone.feature [hunk 12]
```diff
@@ -27,9 +27,99 @@ Feature: cloning an Exosphere application and its services
...
+    Given source control contains the services "web" and "users"
```
Comment (kevgo) [inline features/clone.feature]: Only create "users" here

File: features/steps/steps.ls [hunk 13]
```diff
@@ -38,17 +40,32 @@ module.exports = ->
       @make-repo dir
 
 
-  @When /^running "([^"]*)" in the "([^"]*)" directory$/ (command, directory, done) ->
+  @Given /^the "([^"]*)" clone already exists in the workspace$/ (repo-name, done) ->
+    @process = new ObservableProcess("git clone origins/#{repo-name}",
```
Comment (kevgo) [inline features/steps/steps.ls]: Just create a directory here, using fs.mkdir-sync.

File: features/steps/steps.ls [hunk 14]
```diff
@@ -38,17 +40,32 @@ module.exports = ->
...
+                                     stderr: dim-console.process.stderr)
+      ..on 'ended', (err, exit-code) ->
+        | err  =>  done err
+        | _    =>  done!
+
+
+  @When /^running "([^"]*)" in the terminal$/ (command, done) ->
     @process = new ObservableProcess(call-args(path.join process.cwd!, 'bin', command),
-                                     cwd: path.join(process.cwd!, directory)
+                                     cwd: path.join(process.cwd!, 'tmp')
```
Comment (kevgo) [inline features/steps/steps.ls]: Nice cleanup!

File: features/steps/steps.ls [hunk 15]
```diff
@@ -38,17 +40,32 @@ module.exports = ->
...
                                      stdout: dim-console.process.stdout
                                      stderr: dim-console.process.stderr)
       ..on 'ended', (err, exit-code) ->
-        console.log err
-        console.log exit-code
-        done!
+        | err  =>  done err
```
Comment (kevgo) [inline features/steps/steps.ls]: Thanks for cleaning this up! You can simplify this even more to

```ls
done err
```

If there is no err, this call `done` with a falsy value, which is not considered an error.

File: src/app-cloner.ls [hunk 16]
```diff
@@ -24,7 +24,7 @@ class AppCloner extends EventEmitter
       @app-config = yaml.safe-load fs.read-file-sync(config-path, 'utf8')
       @emit 'app-config-ready', @app-config
       cloners = for service-name in Object.keys @app-config.services
-        service-root = path.join @repository.path, path.dirname @app-config.services[service-name].local
+        service-root = @repository.path
```
Comment (kevgo) [inline src/app-cloner.ls]: What is the motivation behind this change? This no longer makes sense. The root directory of a service should not be the repo path.

File: features/clone.feature [hunk 17]
```diff
@@ -27,9 +27,99 @@ Feature: cloning an Exosphere application and its services
...
-    When running "exo-clone origins/app" in the "tmp" directory
+    When running "exo-clone origins/app" in the terminal
     Then it creates the files:
       | FILE            | FOLDER    |
       | application.yml | app       |
       | service.yml     | app/web   |
       | service.yml     | app/users |
+    Then I am in an empty folder
+
+  Scenario: clone an application with services that reside above application directory.
```
Comment (martinjaime) [inline features/clone.feature]: @kevgo, they were described poorly. The first scenario describes a scenario with services that reside in the app directory. 

The second scenario describes a situation where the cloned service should reside in the same directory as the app directory.

I can see how both sound very similar, but they're testing for different locations of the service clone. Should we merge these into one, where the `users` service is in `./users` and `web` is in `../web`? The reason I wrote this scenario is because this feature did not work before this PR.

File: features/clone.feature [hunk 18]
```diff
@@ -27,9 +27,99 @@ Feature: cloning an Exosphere application and its services
...
+          origin: ../origins/nonexisting
+        web:
+          local: ../web
+          origin: ../origins/web
+      """
+    When running "exo-clone origins/app" in the terminal
+    Then it prints "fatal: repository '../origins/nonexisting' does not exist" in the terminal
+    And it prints "web  done." in the terminal
+    And it prints "exo-clone  Done" in the terminal
+    And it creates the files:
```
Comment (martinjaime) [inline features/clone.feature]: @kevgo, are we sure we want to do this? In a [comment from a previous PR](https://github.com/Originate/exosphere-sdk/pull/92#r72879932) you argued that we should not remove anything so that the user can inspect the results. 

What this scenario describes is that the "app", and "app/web" directories were created, but the "nonexisting" service directory was not created.

I added a step to test that the "app/nonexisting" directory was not created.

File: features/clone.feature [hunk 19]
```diff
@@ -27,9 +27,99 @@ Feature: cloning an Exosphere application and its services
...
+      description: Demonstrates basic Exosphere application startup
+      version: '1.0'
+
+      services:
+        users:
+          local: ./users
+          origin: ../origins/users
+      """
+    And the "app" clone already exists in the workspace
+    When running "exo-clone origins/app" in the terminal
```
Comment (martinjaime) [inline features/clone.feature]: Should we be this specific? Is there a difference between "trying" and simply "running"?

" Do or do not. There is no 'try'." -- Yoda

File: features/clone.feature [hunk 20]
```diff
@@ -27,9 +27,99 @@ Feature: cloning an Exosphere application and its services
...
+    Then it prints "exo-clone  Error: cloning app failed" in the terminal
```
Comment (martinjaime) [inline features/clone.feature]: Wouldn't this step 
> Then I get the error "exo-clone Error: directory 'app' already exists"

be the same as 

> Then it prints "exo-clone Error: directory 'app' already exists" in the terminal

Since logger does not print to `stderr`? We cannot differentiate between an error, and regular output.

How about simply 

> Then it prints "exo-clone  fatal: destination path 'app' already exists and is not an empty directory." in the terminal


since this is the git message that logger outputs to the console.

File: features/clone.feature [hunk 21]
```diff
@@ -27,9 +27,99 @@ Feature: cloning an Exosphere application and its services
...
+      name: Example application
+      description: Demonstrates basic Exosphere application startup
+      version: '1.0'
+
+      services:
+        users:
+          local: ./users
+          origin: ../origins/users
+      """
+    And the "app" clone already exists in the workspace
```
Comment (martinjaime) [inline features/clone.feature]: @kevgo, I took your suggested change, but when I went back to it I remembered why I worded it this way: I want to describe that the "app" clone exists because it is specifically a clone that already exists and not a regular folder so that we can see the `git` failure messages. `git` will not fail if it is a regular folder.

File: src/app-cloner.ls (hunk 16 above)
Comment (martinjaime) [inline src/app-cloner.ls]: That's very true. But I need this path for reference to use the paths specified in service.yml. Should we rename it to something like `reference-path`?

**Comment by martinjaime** [comment]
@kevgo, your comments have been addressed or replied to. Please see https://github.com/Originate/exo-clone/tree/mj-more-tests to see the source code, since I do not see it being updated here.

File: features/clone.feature [hunk 22]
```diff
@@ -27,9 +27,99 @@ Feature: cloning an Exosphere application and its services
...
           origin: ../origins/users
       """
-    When running "exo-clone origins/app" in the "tmp" directory
+    When running "exo-clone origins/app" in the terminal
     Then it creates the files:
       | FILE            | FOLDER    |
       | application.yml | app       |
       | service.yml     | app/web   |
       | service.yml     | app/users |
+    Then I am in an empty folder
```
Comment (kevgo) [inline features/clone.feature]: There should be a cleanup before a scenario runs. I like not doing a cleanup afterwards, so that if a scenario fails, it can be inspected.

File: features/clone.feature (hunk 17 above)
Comment (kevgo) [inline features/clone.feature]: It's better to have separate scenarios for different aspects. But make them super focussed: only one service in an application.

File: features/clone.feature (hunk 18 above)
Comment (kevgo) [inline features/clone.feature]: Tests should not clean up afterwards, so that developers can inspect the mess that their code produces. But the "exo clone" command should not leave non-working code on the hard drive. End users should not have to see and deal with mess.

File: features/clone.feature (hunk 19 above)
Comment (kevgo) [inline features/clone.feature]: The difference is that when I say `When running ...`, then the step should fail if the command was not successful. When I say `When trying to run ...`, then the step should fail if the command succeeds.

File: features/clone.feature (hunk 20 above)
Comment (kevgo) [inline features/clone.feature]: Yes, these steps can behave the same, but I think its important here to say that the user gets an error message, versus just some random output. How we verify the error message programmatically is another question.

File: src/app-cloner.ls (hunk 16 above)
Comment (kevgo) [inline src/app-cloner.ls]: Can you just use `@repository.path` directly, instead of storing that in a variable and then using that variable?

File: features/clone.feature [hunk 23]
```diff
@@ -27,9 +27,107 @@ Feature: cloning an Exosphere application and its services
...
+      description: Demonstrates Exosphere application with service residing above app directory
+      version: '1.0'
+
+      services:
+        web:
+          local: ../web
+          origin: ../origins/web
+      """
+    When running "exo-clone origins/app" in the terminal
+    Then it creates the files:
```
Comment (kevgo) [inline features/clone.feature]: Let's make it easier to read these specs:

```cucumber
Then it creates the files:
  | app/application.yml |
  | web/service.yml     |
```

File: features/clone.feature [hunk 24]
```diff
@@ -27,9 +27,107 @@ Feature: cloning an Exosphere application and its services
...
       | FILE            | FOLDER    |
       | application.yml | app       |
       | service.yml     | app/web   |
       | service.yml     | app/users |
+
+  Scenario: clone an application with services that reside above application directory.
+    Given I am in an empty folder
+    And source control contains a repo "web" with a file "service.yml" and the content:
+      """
+
```
Comment (kevgo) [inline features/clone.feature]: Why an empty content here? If the content of the file doesn't matter, then don't specify it in the test, and let the test pick a reasonable default for you:

```cucumber
And source control contains a "web" service

File: features/clone.feature [hunk 25]
```diff
@@ -27,9 +27,107 @@ Feature: cloning an Exosphere application and its services
...
+          origin: ../origins/nonexisting
+        web:
+          local: ./web
+          origin: ../origins/web
+      """
+    When running "exo-clone origins/app" in the terminal
+    Then it prints "fatal: repository '../origins/nonexisting' does not exist" in the terminal
+    And it prints "web  done." in the terminal
+    And it prints "exo-clone  Failed" in the terminal
+    And "app/nonexisting" directory is absent
```
Comment (kevgo) [inline features/clone.feature]: I think the whole `app` directory should not exist.

File: features/clone.feature [hunk 26]
```diff
@@ -27,9 +27,107 @@ Feature: cloning an Exosphere application and its services
...
+      version: '1.0'
+
+      services:
+        users:
+          local: ./users
+          origin: ../origins/users
+      """
+    And my workspace already contains the folder "app"
+    When running "exo-clone origins/app" in the terminal
+    Then it prints "exo-clone  Error: cloning app failed" in the terminal
```
Comment (kevgo) [inline features/clone.feature]: Tell the user exactly what is going on here: "Error: Cannot clone into an already existing directory"

File: features/clone.feature [hunk 27]
```diff
@@ -27,9 +27,107 @@ Feature: cloning an Exosphere application and its services
...
+    And no new files or folders have been created
+
+
+  Scenario: a service directory that already exists
```
Comment (kevgo) [inline features/clone.feature]: Remove "that'

File: features/clone.feature [hunk 28]
```diff
@@ -27,9 +27,107 @@ Feature: cloning an Exosphere application and its services
...
+
+      services:
+        users:
+          local: ../users
+          origin: ../origins/users
+      """
+    And my workspace already contains the folder "users"
+    When running "exo-clone origins/app" in the terminal
+    Then it prints "users  Service cloning failed" in the terminal
+    And it prints "exo-clone  Some services failed to clone or to be verified." in the terminal
```
Comment (kevgo) [inline features/clone.feature]: Be more specific, please tell the user exactly which service failed to clone. Also, what does "could not be verified" mean here?

File: features/steps/steps.ls [hunk 29]
```diff
@@ -38,17 +45,46 @@ module.exports = ->
       @make-repo dir
 
 
-  @When /^running "([^"]*)" in the "([^"]*)" directory$/ (command, directory, done) ->
+  @Given /^my workspace already contains the folder "([^"]*)"$/ (folder-name) ->
+    existing-folders.add folder-name
+    folder-name = path.join 'tmp', folder-name
+    temp-file = path.join folder-name, 'temp-file'
+    fs.mkdirs-sync folder-name
+    fd = fs.open-sync temp-file, 'w'
```
Comment (kevgo) [inline features/steps/steps.ls]: Just use `mkdirp` to create a file together with its folder in one swoop.

File: features/steps/steps.ls [hunk 30]
```diff
@@ -38,17 +45,46 @@ module.exports = ->
...
+                                     cwd: path.join(process.cwd!, 'tmp')
                                      stdout: dim-console.process.stdout
                                      stderr: dim-console.process.stderr)
       ..on 'ended', (err, exit-code) ->
-        console.log err
-        console.log exit-code
-        done!
+        done err
+
+  @Then /^"([^"]*)" directory is absent$/ (dir-path, done) ->
```
Comment (kevgo) [inline features/steps/steps.ls]: I think we can do everything with the "no new files or folders have been created" step, and don't need this step at all.

File: features/steps/steps.ls [hunk 31]
```diff
@@ -38,17 +45,46 @@ module.exports = ->
...
+      existing-folders.add row.FOLDER.split(path.sep)[0]
+
+
+  @Then /^it prints "([^"]*)" in the terminal$/, (expected-text) ->
+    expect(strip-ansi @process.full-output!).to.contain expected-text
+
+
+  @Then /^no new files or folders have been created$/ (done) ->
+    files-list = new Set fs.readdir-sync path.join process.cwd!, 'tmp'
+    new-files = difference files-list, Array.from existing-folders
```
Comment (kevgo) [inline features/steps/steps.ls]: Using a `Set` here requires ES6. We want to be compatible to Node4. Use [unique](http://www.preludels.com/#unique) here.

File: features/steps/steps.ls [hunk 32]
```diff
@@ -38,17 +45,46 @@ module.exports = ->
...
+    unless new-files.length is 0
```
Comment (kevgo) [inline features/steps/steps.ls]: Avoid stand-alone `unless`, use `if` instead, it is much easier to understand.

```ls
if new-files.length > 0

File: features/clone.feature (hunk 28 above)
Comment (martinjaime) [inline features/clone.feature]: It means that a service was invalid. Maybe we could instead say, `<service> is invalid`? Should this line also repeat what the previous line already mentioned as a summary? This is what the output looks like now: 

screen shot 2016-09-22 at 10 14 08 am

File: features/clone.feature (hunk 19 above)
Comment (martinjaime) [inline features/clone.feature]: That makes a lot of sense, thank you!

**Comment by martinjaime** [comment]
@kevgo, latest comments and replies have been addressed.

File: features/clone.feature [hunk 33]
```diff
@@ -27,9 +27,95 @@ Feature: cloning an Exosphere application and its services
...
           origin: ../origins/users
       """
-    When running "exo-clone origins/app" in the "tmp" directory
+    When running "exo-clone origins/app" in the terminal
     Then it creates the files:
-      | FILE            | FOLDER    |
-      | application.yml | app       |
-      | service.yml     | app/web   |
-      | service.yml     | app/users |
+      | FILE                  |
```
Comment (kevgo) [inline features/clone.feature]: Get rid of the table header. Since the step says "it creates the files:", and there is only one column, its pretty obvious that the table contains filenames.

File: features/clone.feature [hunk 34]
```diff
@@ -27,9 +27,95 @@ Feature: cloning an Exosphere application and its services
...
+      | app/application.yml   |
+      | app/web/service.yml   |
+      | app/users/service.yml |
+
+  Scenario: clone an application with services that reside above application directory.
```
Comment (kevgo) [inline features/clone.feature]: above __the__ application directory

File: features/clone.feature [hunk 35]
```diff
@@ -27,9 +27,95 @@ Feature: cloning an Exosphere application and its services
...
+      version: '1.0'
+
+      services:
+        web:
+          local: ../web
+          origin: ../origins/web
+      """
+    When running "exo-clone origins/app" in the terminal
+    Then it creates the files:
+      | FILE                |
```
Comment (kevgo) [inline features/clone.feature]: Same here, no table header, as my example specified. You can access the raw table data to get all its rows.

File: features/clone.feature [hunk 36]
```diff
@@ -27,9 +27,95 @@ Feature: cloning an Exosphere application and its services
...
+      version: '1.0'
+
+      services:
+        users:
+          local: ./users
+          origin: ../origins/users
+      """
+    And my workspace already contains the folder "app"
+    When trying to run "exo-clone origins/app"
+    Then I get the error "exo-clone  fatal: destination path 'app' already exists and is not an empty directory."
```
Comment (kevgo) [inline features/clone.feature]: Please remove "and is not an empty directory". The error message is expressive enough without it.

> application directory 'app' already exists

File: features/clone.feature [hunk 37]
```diff
@@ -27,9 +27,95 @@ Feature: cloning an Exosphere application and its services
...
+      | web/service.yml     |
+
+
+  Scenario: Cloning an application with a nonexisting service.
+    Given I am in an empty folder
+    And source control contains a "web" service
+    And source control contains a repo "app" with a file "application.yml" and the content:
+      """
+      name: Example application
+      description: Demonstrates basic Exosphere application startup
```
Comment (kevgo) [inline features/clone.feature]: description: contains a non-existing service

File: features/clone.feature [hunk 38]
```diff
@@ -27,9 +27,95 @@ Feature: cloning an Exosphere application and its services
...
+    And it prints "exo-clone  Failed" in the terminal
+    And no new files or folders have been created
+
+  Scenario: the application directory already exists
+    Given I am in an empty folder
+    And source control contains a "users" service
+    And source control contains a repo "app" with a file "application.yml" and the content:
+      """
+      name: Example application
+      description: Demonstrates basic Exosphere application startup
```
Comment (kevgo) [inline features/clone.feature]: description: is going to be cloned into an already existing directory

File: features/clone.feature [hunk 39]
```diff
@@ -27,9 +27,95 @@ Feature: cloning an Exosphere application and its services
...
+    And no new files or folders have been created
+
+
+  Scenario: a service directory already exists
+    Given I am in an empty folder
+    And source control contains a "users" service
+    And source control contains a repo "app" with a file "application.yml" and the content:
+      """
+      name: Example application
+      description: Demonstrates basic Exosphere application startup
```
Comment (kevgo) [inline features/clone.feature]: description: contains a service that would be cloned into an already existing directory

File: features/clone.feature [hunk 40]
```diff
@@ -27,9 +27,95 @@ Feature: cloning an Exosphere application and its services
...
+          local: ../users
+          origin: ../origins/users
+      """
+    And my workspace already contains the folder "users"
+    When running "exo-clone origins/app" in the terminal
+    Then I get the error "users  Service cloning failed"
+    And it prints "exo-clone  Some services failed to clone or were invalid Exosphere services." in the terminal
+    And it prints "exo-clone  Failed" in the terminal
+
+
```
Comment (kevgo) [inline features/clone.feature]: Remove the 2 empty lines at the end here

File: features/clone.feature [hunk 41]
```diff
@@ -27,9 +27,95 @@ Feature: cloning an Exosphere application and its services
...
+      version: '1.0'
+
+      services:
+        users:
+          local: ../users
+          origin: ../origins/users
+      """
+    And my workspace already contains the folder "users"
+    When running "exo-clone origins/app" in the terminal
+    Then I get the error "users  Service cloning failed"
```
Comment (kevgo) [inline features/clone.feature]: Can it print a more useful message here, like "users Service directory '/a/b/c/users' already exists"

File: features/steps/steps.ls [hunk 42]
```diff
@@ -1,54 +1,104 @@
...
+    repo-dir = path.join process.cwd!, 'tmp', 'origins', service-name
+    mkdirp repo-dir, (err) ~>
+      | err  =>  return done err
+      fs.write-file-sync path.join(repo-dir, 'service.yml'), ' '
+      @make-repo repo-dir
+      done!
 
 
   @Given /^source control contains a repo "([^"]*)" with a file "([^"]*)" and the content:$/ (app-name, file-name, file-content) ->
+    @existing-folders = unique @existing-folders.push('origins')
```
Comment (kevgo) [inline features/steps/steps.ls]: `@existing-folders = @existing-folders.push 'origins'  |>  unique`

Comment (kevgo) [inline features/steps/steps.ls]: If you don't need `unique` here, you can simplify this to:

```
@existing-folders .= push 'origins'
```

or, since push mutates the array:

```
@existing-folders.push 'origins'

File: features/clone.feature (hunk 41 above)
Comment (martinjaime) [inline features/clone.feature]: git is already passing a useful error message. The message "users  Service cloning failed" is printed right after the git message. It prints `users  fatal: destination path '/home/mrtin/tmp/exo-clone/tmp/users' already exists and is not an empty directory.`

File: features/steps/steps.ls (hunk 42 above)
Comment (martinjaime) [inline features/steps/steps.ls]: unique was to ensure that no duplicates were created if "source control contains.. " steps are called more than once.

**Comment by martinjaime** [comment]
@kevgo, latest comments addressed.

File: features/clone.feature [hunk 43]
```diff
@@ -27,9 +27,91 @@ Feature: cloning an Exosphere application and its services
...
+        nonexisting:
+          local: ./nonexisting
+          origin: ../origins/nonexisting
+        web:
+          local: ./web
+          origin: ../origins/web
+      """
+    When running "exo-clone origins/app" in the terminal
+    Then I get the error "fatal: repository '../origins/nonexisting' does not exist"
+    And it prints "web  done." in the terminal
```
Comment (kevgo) [inline features/clone.feature]: Please remove the dot from here.

File: features/clone.feature (hunk 41 above)
Comment (kevgo) [inline features/clone.feature]: cool, thanks!

File: features/steps/steps.ls [hunk 44]
```diff
@@ -1,54 +1,104 @@
...
+    @existing-folders = @existing-folders.push('origins') |> unique
     repo-dir = path.join process.cwd!, 'tmp', 'origins', app-name
     fs.mkdirs-sync repo-dir
     fs.write-file-sync path.join(repo-dir, file-name), file-content
     @make-repo repo-dir
 
 
-  @Given /^source control contains the services "([^"]*)" and "([^"]*)"$/ (service1, service2) ->
+  @Given /^source control contains the services "([^"]*)" and "([^"]*)"$/ (service1, service2, done) ->
+    @existing-folders = unique @existing-folders.push('origins')
```
Comment (kevgo) [inline features/steps/steps.ls]: Please format this line consistently with line 36.

File: features/steps/steps.ls [hunk 45]
```diff
@@ -1,54 +1,104 @@
...
     repo-dirs = [ path.join(process.cwd!, 'tmp', 'origins', service1),
                   path.join(process.cwd!, 'tmp', 'origins', service2) ]
     for dir in repo-dirs
-      fs.mkdirs-sync dir
+      mkdirp.sync dir
```
Comment (kevgo) [inline features/steps/steps.ls]: Nice!

File: features/steps/steps.ls [hunk 46]
```diff
@@ -1,54 +1,104 @@
...
       file-name = path.join dir, 'service.yml'
-      fd = fs.open-sync file-name, 'w'
       fs.write-file-sync file-name, ''
-      fs.close-sync(fd)
       @make-repo dir
+    done!
+
+
+  @Given /^my workspace already contains the folder "([^"]*)"$/ (folder-name) ->
+    @existing-folders = unique @existing-folders.push(folder-name)
```
Comment (kevgo) [inline features/steps/steps.ls]: Do you really need all these `unique` calls? Which test creates duplicate folders?

File: features/steps/steps.ls [hunk 47]
```diff
@@ -1,54 +1,104 @@
...
+    folder-name = path.join 'tmp', folder-name
+    temp-file = path.join folder-name, 'temp-file'
+    fs.mkdirs-sync folder-name
+    fd = fs.open-sync temp-file, 'w'
```
Comment (kevgo) [inline features/steps/steps.ls]: Can you just say: `fs.write-file-sync path.join(folder-name, 'temp-file'), ' '` here?

Comment (kevgo) [inline features/steps/steps.ls]: Or even better use https://github.com/AoDev/Filendir to do this all in one line.

File: features/steps/steps.ls [hunk 48]
```diff
@@ -1,54 +1,104 @@
...
+
+  @Then /^it prints "([^"]*)" in the terminal$/, (expected-text) ->
+    expect(strip-ansi @process.full-output!).to.contain expected-text
+
+
+  @Then /^I get the error "([^"]*)"$/ (expected-text) ->
+    expect(strip-ansi @process.full-output!).to.contain expected-text
+
+  @Then /^no new files or folders have been created$/ (done) ->
+    files-list = new Set fs.readdir-sync path.join process.cwd!, 'tmp'
```
Comment (kevgo) [inline features/steps/steps.ls]: Using `Set` breaks compatibility with Node 4. Use `unique` here and simplify this to something like:

```ls
done if fs.readdir-sync path.join(process.cwd!, 'tmp') |> unique |> difference(@existing-folders) > 0
  new Error "New files or folders were created: #{new-files}"
else
  null

File: features/steps/steps.ls (hunk 46 above)
Comment (martinjaime) [inline features/steps/steps.ls]: It's not that tests are creating duplicate folders, but it's that tests are creating folders with the same parents. I was trying to avoid having duplicates in that list of `existing folders`. Should we not care about this?

File: features/steps/steps.ls (hunk 48 above)
Comment (martinjaime) [inline features/steps/steps.ls]: Oops, I forgot to remove that one! 

If we do this in one line, we lose access to `new-files`.

**Comment by martinjaime** [comment]
closes #2
